lib/utils/stream.js

```diff
@@ -12,16 +12,19 @@
 var Stream = function() {
   this.init = function() {
     var listeners = {};
+    var sentinels = {};
     /**
      * Add a listener for a specified event type.
      * @param type {string} the event name
      * @param listener {function} the callback to be invoked when an event of
      * the specified type occurs
      */
-    this.on = function(type, listener) {
+    this.on = function(type, listener, sentinel) {
```

Comment:
The sentinel is used by the pipe function internally. Externally, the `pipe`-related api is exactly the same. We can make it so that the "event" api is the same by creating internal-only event binding functions and exposing them via a two-argument proxy. I feel that may be more work for little benefit.